src/Mvc/Mvc.Analyzers/src/TagHelpersInCodeBlocksAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class TagHelpersInCodeBlocksAnalyzer : DiagnosticAnalyzer
+    {
+        public TagHelpersInCodeBlocksAnalyzer()
+        {
+            TagHelperInCodeBlockDiagnostic = DiagnosticDescriptors.MVC1006_FunctionsContainingTagHelpersMustBeAsyncAndReturnTask;
+            SupportedDiagnostics = ImmutableArray.Create(new[] { TagHelperInCodeBlockDiagnostic });
+        }
+
+        private DiagnosticDescriptor TagHelperInCodeBlockDiagnostic { get; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                var symbolCache = new SymbolCache(context.Compilation);
+
+                if (symbolCache.TagHelperRunnerRunAsyncMethodSymbol == null)
+                {
+                    // No-op if we can't find bits we care about.
+                    return;
+                }
+
+                InitializeWorker(context, symbolCache);
+            });
+        }
+
+        private void InitializeWorker(CompilationStartAnalysisContext context, SymbolCache symbolCache)
+        {
+            context.RegisterOperationBlockStartAction(startBlockContext =>
+            {
+                startBlockContext.RegisterOperationAction(context =>
```

Comment:
It might be useful to check the `startBlockContext` up front to prevent doing a bunch of work on async methods. If that's super hard them I'm not worried about it.

Reply:
I could do it for methods in `@functions` but I feel like that separation of logic would be a little wonky to have separated here. Everything else I couldn't because of how operations work on nested methods aka
```
@{
    void Foo1()
    {
        <p>Hello WOrld</p>

        Func<Task> func2 = () =>
        {
            <distributed-cache name="foo"></distributed-cache>
        };
    }
}
```